### ISSUE TITLE
Unstable loss during training

### ISSUE THREAD

**Issue by donand** [NVIDIA/waveglow#155]
Hello everyone,

I'm trying to train the waveglow model on a different language dataset, but I'm facing an issue of exploding loss.
The loss keeps decreasing steadily at the beginning, but at a certain point it explodes going from something like -6 to -2/-4, making the output audio just noise.
I tried both starting the training from scratch and starting from the checkpoint provided by NVIDIA (waveglow_256channels.pt) but I face the same issue in all the cases.

This is the loss of the training started from the NVIDIA checkpoint
![image](https://user-images.githubusercontent.com/10143667/64524692-61c05980-d2ff-11e9-99ce-e7b79dcf9efa.png)

This is the loss of the training started from scratch
![image](https://user-images.githubusercontent.com/10143667/64524653-49e8d580-d2ff-11e9-9dbb-0c3dc5701bd3.png)
Here at the end you can see a second spike.

The setup of my experiments is:
- Single V100 GPU
- batch_size: 24 (using gradient accumulation, single batch is 6 but it's accumulated over 4 steps)
- learning_rate: 1e-4 (tried also 5e-5, 2e-5, 1e-5)

I tried several things, but nothing worked:
- Reduced the learning rate till 1e-5 to try to make the model more stable
- Increased the batch size to 24, before it was 12
- I tried to clip the gradients to make the model more stable, but it didn't work

Is anyone facing the same problem? Any insight?

Thank you very much

**Comment by jongwook** [comment]
I'm also seeing the problem, although not as sever as your graph (about +1.5 jump in loss). I'd appreciate if the authors can share some techniques they used to avoid this!

**Comment by rafaelvalle** [comment]
Make sure your samples are at least as long as the segment length in the config otherwise they will be zero padded and this can cause instability in optimization such as the one you are seeing.

**Comment by jongwook** [comment]
I'm testing with the [MAESTRO v2] dataset which are significantly longer, but still seeing the jump. I'm actually chunking the inputs to shorter segments to help with data loading.

I'll try doing rejection sampling to exclude silence input, to see if it helps.

**Comment by donand** [comment]
I checked the whole dataset and all the samples are longer than segment_length, so they're not being padded. I also tried rejection sampling to avoid silent segments but I never found a silent segment in training.

I ran a new experiment after the previous two checks and I faced the same issue, so these are not the problems that are causing the spike in the loss.
At this point I don't really know what could be the reason.

**Comment by hadaev8** [comment]
16k segment_length mean 16 secs?

**Comment by jongwook** [comment]
Just another bit of clue, with the default 80-bin Mel spectrogram the implementation in this repo ran fine, but when I tried 128-bin, I had to avoid using fp16 optimization. It trains slower than fp16, but it doesn't suddenly jump +3 like the tensorboard graph above.

**Comment by rafaelvalle** [comment]
@donand are you using the default config in the repo?

**Comment by donand** [comment]
@rafaelvalle For the most part yes, but I made some changes, like matching the sampling rate of the dataset (16k), trying to decrease the learning rate to make the training stable, reducing the batch size to fit in the GPU and I added the gradient accumulation to accumulate gradients over steps to increase the effective batch size. The other fields should be unchanged.

This is my config.json
`{
    "train_config": {
        "fp16_run": false,
        "output_directory": "checkpoints",
        "epochs": 100000,
        "learning_rate": 5e-5,
        "sigma": 1.0,
        "iters_per_checkpoint": 1000,
        "batch_size": 6,
        "seed": 1234,
        "checkpoint_path": "checkpoints/waveglow_0",
        "steps_acc_grad": 4,
        "log_directory": "logs"
    },
    "data_config": {
        "training_files": "train_files.txt",
        "segment_length": 16000,
        "sampling_rate": 16000,
        "filter_length": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "mel_fmin": 0.0,
        "mel_fmax": 8000.0
    },
    "dist_config": {
        "dist_backend": "nccl",
        "dist_url": "tcp://localhost:54321"
    },

    "waveglow_config": {
        "n_mel_channels": 80,
        "n_flows": 12,
        "n_group": 8,
        "n_early_every": 4,
        "n_early_size": 2,
        "WN_config": {
            "n_layers": 8,
            "n_channels": 256,
            "kernel_size": 3
        }
    }
}`

**Comment by Taekyoon** [comment]
@jongwook I have a question. How did you set `while audio_std < 1e-5:` to avoid silent sounds?

**Comment by jongwook** [comment]
@Taekyoon I thought the code above is quite straightforward; `segment.std()` being smaller than `1e-5` means that the audio segment is practically silent.

**Comment by cduguet** [comment]
It is important to notice that this code only checks for silence in case of audios longer than sequence length. You should still reject shorter sequences or pad them with noise. 

> Rejection sampling to avoid silent training samples seems to help in my case.
> 
> In `mel2samp.py`:
> 
> ```python
>         # Take segment
>         if audio.size(0) >= self.segment_length:
>             audio_std = 0
>             while audio_std < 1e-5:
>                 max_audio_start = audio.size(0) - self.segment_length
>                 audio_start = random.randint(0, max_audio_start)
>                 segment = audio[audio_start:audio_start+self.segment_length]
>                 audio_std = segment.std()
>             audio = segment
> ```

**Comment by puppyapple** [comment]
hello guys, I tried to train the model with my dataset. I've ensured all audio samples has length larger than segment length and use the sdt() trick advised by @jongwook  to avoid silent subsequence but this still gives me an explode of loss after only 50 epochs. Any further ideas for this?

**Comment by ishihara1989** [comment]
I have currently no problem despite that my dataset contains some sample whose length is less than segment_length.
Not sure but following modification might help.

1. set `segment_length=16384` so that segment_length is multiple of STFT's hop_length(=256)
2. use dequantization before forward processing, which looks like: 
  `audio = audio + (torch.rand_like(audio)-0.5)/32768.0`

**Comment by puppyapple** [comment]
@ishihara1989 thanks for sharing, I will give it a try :)

**Comment by tdplaza** [comment]
> I have currently no problem despite that my dataset contains some sample whose length is less than segment_length.
> Not sure but following modification might help.
> 
> 1. set `segment_length=16384` so that segment_length is multiple of STFT's hop_length(=256)
> 2. use dequantization before forward processing, which looks like:
>    `audio = audio + (torch.rand_like(audio)-0.5)/32768.0`

@ishihara1989 Could you please tell me where to put this dequantization? Thank you so much.

**Comment by CookiePPP** [comment]
@tdplaza 
If you implement it and it works, please share!

Also, you could probably add it
https://github.com/NVIDIA/waveglow/blob/master/mel2samp.py#L103
```
        audio = audio / MAX_WAV_VALUE
```
just after this line. To look like
```
        audio = audio / MAX_WAV_VALUE
        audio = audio + (torch.rand_like(audio)-0.5)/MAX_WAV_VALUE
```

**Comment by MuyangDu** [comment]
> @tdplaza
> If you implement it and it improves quality, please share!
> 
> Also, you could probably add it
> https://github.com/NVIDIA/waveglow/blob/master/mel2samp.py#L103
> 
> ```
>         audio = audio / MAX_WAV_VALUE
> ```
> 
> just before this line. To look like
> 
> ```
>         audio = audio + (torch.rand_like(audio)-0.5)/MAX_WAV_VALUE
>         audio = audio / MAX_WAV_VALUE
> ```

I have tried this. the loss still become NaN after several epochs.

**Comment by rafaelvalle** [comment]
For the people getting NaNs after several epochs: please check if you have z values equal to 0 or log_s values that are NaNs.

**Comment by raikarsagar** [comment]
I am also facing same issue of NaNs after 20k steps, log_s values are -inf and loss is NaN, 
Any solutions?

**Comment by KB00100100** [comment]
I also use Single V100 GPU to train my model, but it will take hundreds of days to finish 100000 epochs. I wonder to ask how many hours you guys taken for training yourself models.
